traits/src/data_provider.rs

```diff
@@ -0,0 +1,138 @@
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+/// Data provider with ability to provide data with no-op, and provide all data.
+pub trait DataFeeder<Key, Value, AccountId>: DataProvider<Key, Value> {
+	/// Provide a new value for a given key from an operator
+	fn feed_value(who: AccountId, key: Key, value: Value) -> DispatchResult;
+}
+
+/// A simple trait to provide data
+pub trait DataProvider<Key, Value> {
+	/// Get data by key
+	fn get(key: &Key) -> Option<Value>;
+}
+
+/// Extended data provider to provide timestamped data by key with no-op, and
+/// all data.
+pub trait DataProviderExtended<Key, TimestampedValue> {
+	/// Get timestamped value by key
+	fn get_no_op(key: &Key) -> Option<TimestampedValue>;
+	/// Provide a list of tuples of key and timestamped value
+	fn get_all_values() -> Vec<(Key, Option<TimestampedValue>)>;
+}
+
+#[allow(dead_code)] // rust cannot defect usage in macro_rules
+pub fn median<T: Ord + Clone>(mut items: Vec<T>) -> Option<T> {
+	if items.is_empty() {
+		return None;
+	}
+
+	let mid_index = items.len() / 2;
+	// Won't panic as guarded items not empty case.
+	let (_, median, _) = items.partition_at_index(mid_index);
+	Some(median.clone())
+}
+
+#[macro_export]
+macro_rules! create_median_value_data_provider {
+	($name:ident, $key:ty, $value:ty, $timestamped_value:ty, [$( $provider:ty ),*]) => {
+		pub struct $name;
+		impl $crate::DataProvider<$key, $value> for $name {
+			fn get(key: &$key) -> Option<$value> {
+				let mut values = vec![];
+				$(
+					if let Some(v) = <$provider as $crate::DataProvider<$key, $value>>::get(&key) {
+						values.push(v);
+					}
+				)*
+				$crate::data_provider::median(values)
+			}
+		}
+		impl $crate::DataProviderExtended<$key, $timestamped_value> for $name {
+			fn get_no_op(key: &$key) -> Option<$timestamped_value> {
+				let mut values = vec![];
+				$(
+					if let Some(v) = <$provider as $crate::DataProviderExtended<$key, $timestamped_value>>::get_no_op(&key) {
+						values.push(v);
+					}
+				)*
+				$crate::data_provider::median(values)
+			}
+			fn get_all_values() -> Vec<($key, Option<$timestamped_value>)> {
+				let mut keys = sp_std::collections::btree_set::BTreeSet::new();
+				$(
+					<$provider as $crate::DataProviderExtended<$key, $timestamped_value>>::get_all_values()
+						.into_iter()
+						.for_each(|(k, _)| { keys.insert(k); });
```

Comment:
not sure if `.map(|(k, _) k|).collect::<BTreeSet>::()` will work

Reply:
Hmm, I tried, it works but doesn't read as straightforward as the previous way IMO, and also had to introduce an extra vec:

```rust
let keys = vec![$( <$provider as $crate::DataProviderExtended<$key, $timestamped_value>>::get_all_values() ),*]
    .into_iter()
    .flatten()
    .map(|(k, _)| k).collect::<sp_std::collections::btree_set::BTreeSet<_>>();
```